src/Elastic.Apm.AspNetCore/ApmMiddleware.cs

```diff
@@ -42,21 +42,37 @@ public ApmMiddleware(RequestDelegate next, Tracer tracer, IApmAgent agent)
 
 		public async Task InvokeAsync(HttpContext context)
 		{
-			var transaction = StartTransaction(context);
+			var transaction = StartTransactionAsync(context);
+			await FillSampledTransactionContextRequest(transaction, context);
 
 			try
 			{
 				await _next(context);
 			}
-			catch (Exception e) when (transaction != null && ExceptionFilter.Capture(e, transaction, context, _logger))
-			{ }
+			catch (Exception e) when (transaction != null)
+			{
+				transaction.CaptureException(e);
+				// It'd be nice to have this in an exception filter, but that would force us capturing the request body synchronously.
+				// Therefore we rather unwind the stack in the catch block and call the async method.
+				if (context != null && transaction.IsContextCreated)

```

Comment:
`transaction.IsContextCreated` is redundant since `CollectRequestBody` includes it.

Reply:
Removed. This property confuses me :)